### ISSUE TITLE
Captcha always invalid on first submit

### ISSUE THREAD

**Issue by dev-rke** [maikschneider/bw_captcha#18]
**Description**
The captcha is always the same when visiting a page, and fails on submit.
It seems the captcha is bound to the page (and is therefore cached), instead it is bound to the user.
With the current implementation, this extension cannot be used.

**Steps to reproduce**
To test the behaviour, create a new form page, add a captcha element.
Open a first browser in private mode (e.g. Firefox) and open the form.
Open a second browser in private mode (e.g. Chrome) and open the form.

You will see that the captcha in both browser instances are the same.
This also guarantees that the first submit of the captcha is ALWAYS invalid.

**Environment**
TYPO3: 11.5.23
bw_captcha: 2.0.4
Our production environment uses an azure gateway and we are hosting using nginx.

The behaviour can be confirmed also on my local development environment, i also tested with bw_captcha 2.0.2 and 2.0.3. 
Furthermore i did a test using TYPO3 11.5.22, no difference.

**Ideas**
I like the extension and would like to support to solve this issue.
I used xdebug and placed a call within your FormElementCaptchaHook class.
My first observation is, that the beforeRendering hook will only be rendered once, until the form is cached.
Is that expected?

One idea might be to implement the image not inline, but do a call to a custom middleware, which always generates a fresh captcha image and sets a private / nocache http header. I think that might solve all caching issues.

Thanks for that great extension, let's make it greater! :-)

**Comment by dev-rke** [comment]
I just developed a solution. Unfortunately there seems to be no valid approach to render a form element uncached via the form extension.
Therefore i decided to render an uncached content object within the form element which itself just renders the relevant part of the form.

I created a new typoscript lib
```typoscript
lib.bw_captcha.captcha = COA_INT
lib.bw_captcha.captcha {
    10 = FLUIDTEMPLATE
    10.dataProcessing.10 = MyVendor\MyCompany\DataProcessing\FormCaptchaProcessor
    10.file = EXT:myextension/Resources/Private/Templates/form/Frontend/bw_captcha.html
}
```

This lib is triggered by an overriden captcha form element:
```html
<div class="captcha">
    {f:cObject(typoscriptObjectPath: 'lib.bw_captcha.captcha')}
</div>
```

I migrated your FormCaptchaHook to a FormCaptchaProcessor and instead of applying properties on $renderable i just return an array with the same settings. These settings are applied to the image and the reload link in bw_captcha.html template.

Locally it works like a charm.
The typoscript lib is uncached and will render on every frontend call the fluid template and therefore the dataprocessor will run.

I tested the following behaviour successfully:

- Render the form and see a captcha
- Submit the correct filled form, captcha is validated successfully
- Submit an incorrect filled captcha, captcha will notify the user, that the solved captcha was incorrect
- Trigger captcha reload button, see another captcha and successfully submit the form

Do you see anything that might cause issues?
Are you interested in a pull request?

**Comment by maikschneider** [comment]
Hi @dev-rke, thanks for your feedback and ideas! 

Have you looked into #11 ? Over there I described, how to generate a new image at every page refresh. Could you please test the JavaScript solution if this resolves your problem?

The main "problem" is the cached captcha. Like I stated in the the linked issue, I'm not a big fan of breaking the whole page cache just because of an included form with a captcha element. If you really want a fresh captcha without JS, disable the cache of the page. Your solution with `COA_INT` does exactly this: Disabling page cache generation. These elements can be inspected via admin panel and are bad practice IMHO. 

Your setup with an azure gateway sounds very interesting. I'm not sure if this has any consequences on session handling (captcha ID is handled in frontend user session). I'll have a look into this next week. In the meantime, please test and report again with the suggested solution.

**Comment by dev-rke** [comment]
Hi @maikschneider ,

thanks for your quick response. :-) 
I read #17 before digging into the code. I understood that the bug described there only happens **after** the first submit.
My case happens **always**, even before the first submit, as the captcha is the same for different persons on different IPs!
That azure gateway is configured to pipe anything through, no caching enabled, it just looks for malicious requests.
The bug itself can reproduced without the gateway on a local development environment.

Therefore i decided to implement a valid and stable solution based upon core functionality.
I do not like JavaScript based solutions, as the page should work always and also with disabled or defective javascript (one bug kills all), especially on form elements and should avoid unnecessary requests.
Including a captcha is a valid reason to invalidate a section cache, as most parts of the page are still being cached.
In general: even with a cached page TYPO3 renders your page with cache headers `Cache-Control: private, no store` which indicates that browser caching is disabled. The reason for this is that the page might change under the same url, as the plugin will generate post requests.

Using `COA_INT` isn't bad practice at all. It is bad practice to disable page caching in general like using `no_cache=1` or using `COA_INT` with slow code sections.
My implementation only uncaches the html section of the image, so the rendering of all the form elements is still cached.
In my case the rendering times cached are around 70ms, with the COA_INT it degrades to 100ms to 120ms.
When rendering the complete form in an uncached mode (flushing frontend caches from backend), it takes up to 3500ms.

I implemented a whole website using COA_INT elements to immediatly update a navigation when the central navigation cache changes. The average rendering times are just around 20ms higher than without that COA_INT on all pages.

I would like to know if you are interested in a pull request, then i will reorganize my code into your extension and wait for your feedback and hope for the next release.

As alternative: what do you think about pointing the img src to the output of a psr15 middleware? That middleware also gets all cookies, so the session is valid on image generation and can be used to access the session and caching objects. The image itself will be always delivered uncached, so browsers always create a new request.

Thank you and have nice evening.

**Comment by akiessling** [comment]
I'd favor a middleware for the image generation. Meanwhile i configured the forms to render as USER_INT as a workaround :/

**Comment by dev-rke** [comment]
@akiessling uargh, that means you have rendering times around 2-3 seconds? Better extract the captcha template as i described above.
Do you see any issues that might come up by creating a middleware? 
E.g. circumventing the captcha? Or generating very much valid captchas, so that it can be easily solved by OCR software?

**Comment by akiessling** [comment]
In my local ddev environment i do have 2-3 seconds, Production should be faster (not yet deployed). Unfortunately, my ticket was not given that much time to work on, so USER_INT was the shortcut. I'll let my customer know that there is a workaround if it's too slow. 

The good old https://extensions.typo3.org/extension/captcha used eID for the image generation, but i'd go for an own middleware if you need to use the Frontenduser session data.
Regarding OCR etc.: i wonder how other CMS / Frameworks generate captchas and tackle such issues. I have not digged into that area, so i have no idea how spammers work and analyze the captchas

**Comment by maikschneider** [comment]
Hey, thanks for your feedback!

Regarding the `COA_INT` cache thing: I rely on [StaticFileCache](https://extensions.typo3.org/extension/staticfilecache) - these elements would prevent the use of this. Since I know quite a few project that use both extensions, `COA_INT` is no option.

I really like the idea of a middleware! In fact I think a dedicated doktype and an uncached plugin action that returns the image would be sufficient. The image source could be included as lazy load, so it's not blocking. Cool, this could finally end the discussion about not refreshed images.

 I'll create an implementation next week.

**Comment by akiessling** [comment]
COA_INT was not a suggestion, just my "quickfix" ;) 

Ping me if you need someone to test! doktype has the benefit (and drawback) of loading the TS --> might be slower, but we can continue to use the TypoScript config. Having the middleware after pretty much everything is initialized probably won't bring a big advantage in terms of loading speed anyway

**Comment by dev-rke** [comment]
Hi @maikschneider ,

static file cache is something i keep my hands off. ;-)
I prefer to use the caching framework of TYPO3 itself, as alternative nginx caching and in rare cases Varnish.
I understand, that COA_INT might generate issues regarding full page caching, but i assume your form won't be cached already via static file cache, as the plugin sends the `Cache-Control: private, no-store` header by default. Any proxies are informed to not cache this kind of request. 
Even if it is going to be cached, you are creating a user session. 
When you have a user session cookie, typically caches are configured to pass through to the TYPO3 frontend.

Actually i think there is only a low chance any form is cached by default. :D 

In general i'd prefer a middleware approach as well, as it is easier to maintain and reduces the amount of TYPO3 that has to be initialized (propably only routing and user session).

**Comment by dev-rke** [comment]
Hi @maikschneider and @akiessling ,

just want to let you know:
i refactored the whole extension tonight.
I will intensify my tests tomorrow. At the moment it looks great. When looking at myself, i am looking tired. :D 
Unfortunately there is not much left of the original extension, so i am not sure, if you are willing to merge the new concept.

Something i'd like to discuss: i am using bootstrap 5 and therefore i depend on FrontendVersion2 of the form extension.
What bootstrap versions are you using? My thoughts are to migrate to a bootstrap 5 compatible template if you agree.

Commit Message as sneak peak:
```
Major refactoring:
+ migrated controllers to captcha middleware
+ removed hook
+ dropped cache integration
+ use user session only for storing captcha phrases
+ added captcha lifetime option to typoscript
+ implemented new refresh button logic
```

**Comment by dev-rke** [comment]
Hi @maikschneider ,
i know everybody has a lot to do, but i'd like kindly ask if you already were able to have a look on my pull request? #21

**Comment by EvilBMP** [comment]
@dev-rke I added some comments to the pull request.

**Comment by maikschneider** [comment]
@dev-rke As mentioned in the pull request, the dependency injection of `TypoScriptFrontendController` needs to be replaced in order to work in v12.

**Comment by maikschneider** [comment]
Solved in [v3.0.0](https://github.com/maikschneider/bw_captcha/releases/tag/v3.0.0)

**Comment by Geddo** [comment]
If you're using static_filecache, you'll need to exclude `.captcha` paths in your .htaccess file to ensure captchas work correctly for external/non-authenticated users:

```
# We only redirect GET requests
RewriteCond %{REQUEST_METHOD} GET

# Exclude .captcha paths from caching
RewriteCond %{REQUEST_URI} !\.captcha [NC]

# Rewrite the request to the static file
RewriteRule .* %{ENV:SFC_ROOT}/%{ENV:SFC_FULLPATH} [L]
```

This prevents caching issues with captcha images for external/non-logged-in users (SFC is mostly disabled for logged in users).

Background:
When refreshing the captcha, the request includes a timestamp parameter (e.g. .captcha?now=1736332376649) which naturally bypasses the static file cache. However, on the initial load, this parameter is missing, causing the static file cache to serve a cached version instead of generating a fresh captcha image. Adding the exclusion rule ensures captchas work correctly on first load.
